src/redis_lib/mod.rs

```diff
@@ -10,44 +11,134 @@ use tracing::error;
+    let redis_port: Result<u16, _> = env::var("REDIS_PORT").unwrap_or("6379".to_owned()).parse();
+    if let Err(err) = redis_port {
+        error!(error = err.to_string(), "Failed to parse int");
+        return None;
+    }
+    let redis_port = redis_port.unwrap();
+
+    let redis_hostname = env::var("REDIS_HOST").unwrap_or("127.0.0.1".to_owned());
+
+    let client = Client::open(format!("redis://{redis_hostname}:{redis_port}"))
+        .map_err(|e| {
+            error!(error = e.to_string(), "Problem creating redis client");
+        })
+        .ok()?;
+
+    let connection = match client.get_async_connection().await {
         Err(err) => {
-            error!(error=err.to_string(), "Problem creating redis client");
+            error!(error = err.to_string(), "Failed connecting to Redis");
             return None;
         }
+        Ok(con) => con,
+    };
+
+    Some((client, connection))
+}
+
+/// Increments the counter that tracks how often an email was successfully used to verify an account
+#[allow(unused)]
+pub async fn increment_email_used_verification_success(con: &mut Connection, email: &str) {
+    let email_hash = hash_email(email);
+    let current = get_email_stats(con, &email_hash).await;
+    set_email_stats(con, &email_hash, current.0 + 1, current.1);
+}
+
+/// Increments the counter that tracks how often an email was given to verify an account
+#[allow(unused)]
+pub async fn increment_email_used_verification(con: &mut Connection, email: &str) {
+    let email_hash = hash_email(email);
+    let current = get_email_stats(con, &email_hash).await;
+    set_email_stats(con, &email_hash, current.0, current.1 + 1);
+}
+
+/// gets how often an email was used to verify an account (not necessarily successful)
+#[allow(unused)]
+pub async fn get_email_used_verification_success(con: &mut Connection, email: &str) -> u64 {
+    get_email_stats(con, &hash_email(email)).await.0
+}
+
+/// gets how often an email was used to verify an account (not necessarily successful)
+#[allow(unused)]
+pub async fn get_email_used_verification(con: &mut Connection, email: &str) -> u64 {
+    get_email_stats(con, &hash_email(email)).await.1
+}
+
+async fn set_email_stats(con: &mut Connection, email_hash: &str, successful: u64, used: u64) {
+    let command: Result<(), _> = con.set(email_hash, format!("{successful},{used}")).await;
+    if let Err(err) = command {
+        error!(
+            error = err.to_string(),
+            "Could not set stats for email {email_hash}, ({successful},{used})"
+        );
     }
-    if let Err(err) = client.get_connection() {
-        error!(error=err.to_string(), "Failed connecting to Redis");
-        return None;
+}
+
+async fn get_email_stats(con: &mut Connection, email_hash: &str) -> (u64, u64) {
+    let command: Result<Option<String>, _> = con.get(email_hash).await;
+    if let Err(err) = command {
+        error!(error = err.to_string(), "Could not read email stats");
+        return (0, 0);
+    }
```

Comment:
This is a different problem then an email address not being in redis. This indicates a problem with the redis connection. Return a different Type so that the calling function knows that redis has a problem. The Method could for example return an `Option<(u64, u64)>`